### PR TITLE
[FLINK-20020][client] Make UnsuccessfulExecutionException part of the JobClient.getJobExecutionResult() contract

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Currently, different implementations of the `JobClient` throw different exceptions. The `ClusterClientJobClientAdapter` wraps the exception from the `JobResult.toJobExecutionResult()` into a `ProgramInvocationException`, the `MiniClusterJobClient` simply wraps it in a `CompletionException` and the `EmbeddedJobClient` wraps it into an `UnsuccessfulExecutionException`. `JobResult.toJobExecutionResult()` should make the exception uniform and part of the contract through `JobExecutionResultException` which exception includes the status of the application.*

## Brief change log

  - *Add `JobExecutionResultException` to make the exception uniform and part of the contract for `JobResult.toJobExecutionResult()`.*
  - *`ClusterClientJobClientAdapter`, `MiniClusterJobClient` and `EmbeddedJobClient` uniform the exception to `JobExecutionResultException`.*

## Verifying this change

  - *Modify `ApplicationDispatcherBootstrapTest` and `PerJobMiniClusterFactoryTest.java` to update the `UnsuccessfulExecutionException` check to the `JobExecutionResultException`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by SteNicholas** [comment]
@kl0u , could you please help to review this commit if you are available?

**Comment by kl0u** [comment]
Hi @SteNicholas , I reviewing your PR now. One thing to add is some tests (for each client) that verify that all clients throw the expected type of exception. This is to make sure that no one breaks the contract in the future. Also this can become part of the javadoc of the `JobClient.getJobExecutionResult()`.

**Comment by SteNicholas** [comment]
> Hi @SteNicholas , I reviewing your PR now. One thing to add is some tests (for each client) that verify that all clients throw the expected type of exception. This is to make sure that no one breaks the contract in the future. Also this can become part of the javadoc of the `JobClient.getJobExecutionResult()`.

I forgot to add `JobExecutionResultException` contract to the javadoc of `JobClient.getJobExecutionResult()`. And about the tests, I thought that current test cases include the tests that verify that all clients throw the expected type of exception. But it's better to add tests for each client extrally.

**Comment by kl0u** [comment]
Also why not putting the logic directly in the `JobResult.toJobExecutionResult()`, instead of keeping it separately to each client? This will lead to more robust code.

**Comment by kl0u** [comment]
Hi @SteNicholas , I have a commit with my comments here https://github.com/apache/flink/tree/client-exception. Also for the tests I think we should just add tests for all the cases in the `JobResultTest`. What do you think?

**Comment by SteNicholas** [comment]
> Hi @SteNicholas , I have a commit with my comments here https://github.com/apache/flink/tree/client-exception. Also for the tests I think we should just add tests for all the cases in the `JobResultTest`. What do you think?

@kl0u Your implementation is better for me than my bad implementation. And the test cases look clear for each client.

**Comment by kl0u** [comment]
Thanks a lot @SteNicholas. Feel free to take the commit and integrate it in your PR :)

**Comment by SteNicholas** [comment]
> Thanks a lot @SteNicholas. Feel free to take the commit and integrate it in your PR :)

@kl0u Thanks for you better implementation commit. I have integrated into my pull request. Please help to review again.

**Comment by kl0u** [comment]
I am having a look right now @SteNicholas

**Comment by SteNicholas** [comment]
> I am having a look right now @SteNicholas 

@kl0u Any other comments by you?

**Review by kl0u — COMMENTED** [review]
@SteNicholas why not putting the `status` in the `JobExecutionException` and remove the `JobExecutionResultException`?

**Comment by SteNicholas** [comment]
> @SteNicholas why not putting the `status` in the `JobExecutionException` and remove the `JobExecutionResultException`?

Because of the compatibility for the previous `JobExecutionException`. I have pushed commit that add the `status` to `JobExecutionException` to check wheter the commit could work. Please review this again.

**Comment by kl0u** [comment]
@SteNicholas I had a look at this solution and I would like to think on how to make the `status` `final`. This means that I am trying to see if we can have all the `JobExecutionExceptions` with a `status`. I will get back with more comments tomorrow.

**Comment by SteNicholas** [comment]
> @SteNicholas I had a look at this solution and I would like to think on how to make the `status` `final`. This means that I am trying to see if we can have all the `JobExecutionExceptions` with a `status`. I will get back with more comments tomorrow.

As you mentioned, all the `JobExecutionExceptions` could be with a given `status`. I have tried to make `status` final and updated `JobCancellationException`, `JobInitializationException` and `JobSubmissionException`. IMO, the `status` of `JobInitializationException` and `JobSubmissionException` is UNKNOWN, and the `status` of `JobCancellationException` is CANCELLED. Please check this.

**Comment by kl0u** [comment]
thanks @SteNicholas , I will check your PR now

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobResult.java [hunk 1]
```diff
@@ -133,26 +132,32 @@ public JobExecutionResult toJobExecutionResult(ClassLoader classLoader) throws J
 					accumulatorResults,
 					classLoader));
 		} else {
-			final Throwable cause;
```
Comment (kl0u) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobResult.java:136]: This `else()` block can now become simply:
`throw getJobExecutionException(serializedThrowable, classLoader);` right? We do not need to wrap the original exception anymore.

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java [hunk 2]
```diff
@@ -669,13 +669,14 @@ public JobExecutionResult executeJobBlocking(JobGraph job) throws JobExecutionEx
...
-			throw new JobExecutionException(job.getJobID(), "Could not retrieve JobResult.", ExceptionUtils.stripExecutionException(e));
+			throw new JobExecutionException(job.getJobID(), ApplicationStatus.UNKNOWN,
+					"Could not retrieve JobResult.", ExceptionUtils.stripExecutionException(e));
 		}
 
 		try {
 			return jobResult.toJobExecutionResult(Thread.currentThread().getContextClassLoader());
 		} catch (IOException | ClassNotFoundException e) {
-			throw new JobExecutionException(job.getJobID(), e);
+			throw new JobExecutionException(job.getJobID(), jobResult.getApplicationStatus(), e);
```
Comment (kl0u) [inline flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:679]: Can this be misleading? For example the application was successful but it failed to deserialize the result so now we have status `SUCCEEDED` but with an exception.

**Review by kl0u — CHANGES_REQUESTED** [review]
@SteNicholas I left some comments in the PR. Please have a look and let me know what you think.

File: flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java [hunk 3]
```diff
@@ -19,41 +19,53 @@
...
+	 * @param cause  The cause of the exception
 	 */
-	public JobExecutionException(JobID jobID, String msg, Throwable cause) {
+	public JobExecutionException(JobID jobID, ApplicationStatus status, String msg, Throwable cause) {
 		super(msg, cause);
 		this.jobID = jobID;
+		this.status = status;
 	}
 
-	public JobExecutionException(JobID jobID, String msg) {
```
Comment (kl0u) [inline flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java:46]: Javadoc.

File: flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java [hunk 4]
```diff
@@ -19,41 +19,53 @@
...
+	public JobExecutionException(JobID jobID, ApplicationStatus status, String msg) {
 		super(msg);
 		this.jobID = jobID;
+		this.status = status;
 	}
 
-	public JobExecutionException(JobID jobID, Throwable cause) {
+	public JobExecutionException(JobID jobID, ApplicationStatus status, Throwable cause) {
```
Comment (kl0u) [inline flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java:58]: Javadoc.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobResult.java (hunk 1 above)
Comment (SteNicholas) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobResult.java:136]: @kl0u Yes, your point is right for me. I forgot to remove the unnecessary wrapper for `JobExecutionException`. I would like to remove this wrapper.

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java (hunk 2 above)
Comment (SteNicholas) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:679]: @kl0u IMO, I don't think this is misleading . The application status for this exception case is SUCCESS, but the exception means the deserialization of job result with an exception. We may add exception message to tell users that the application is succeeded but the deserialization with an exception. What do you think?

**Comment by SteNicholas** [comment]
@tillrohrmann , could you please review this pull request and take a look at the [commit](https://github.com/kl0u/flink/tree/client-exception-fin) of @kl0u ?

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java (hunk 2 above)
Comment (kl0u) [reply flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java]: Maybe this is a nice middle ground @SteNicholas . Have the status as `SUCCESS` and a message that says that something went wrong during getting the result. As a sidenote, it may be that this could require a new type of exception in the future (not necessarily in this PR) as it seems that we are using the same exception for different purposes. One is successful job execution (or not) and the other is successful job result translation (or not). 

This was also a discussion we had offline with @SteNicholas ;)

**Comment by SteNicholas** [comment]
@tillrohrmann , could you please review this pull request for the application status of `JobExecutionException?

**Review by kl0u — COMMENTED** [review]
From a quick look on the final changes, I think your PR looks good now. But I would like also someone else (maybe @tillrohrmann or @aljoscha ) to also have a look about the big picture and how it looks from a user's perspective.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 5]
```diff
@@ -257,7 +258,7 @@ public void run() {
 		runner.join();
 
 		final Throwable t = errorRef.get();
-		if (t != null) {
+		if (t != null && ((JobExecutionException) t).getStatus() == ApplicationStatus.FAILED) {
```
Comment (tillrohrmann) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:261]: Why is this change necessary?

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 6]
```diff
@@ -342,7 +343,7 @@ public void run() {
...
```
Comment (tillrohrmann) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:346]: Same here?

File: flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java [hunk 7]
```diff
@@ -19,41 +19,67 @@
...
+	 * @param jobID  The job's ID.
+	 * @param status The application status.
+	 * @param msg    The cause for the execution exception.
+	 * @param cause  The cause of the exception
 	 */
-	public JobExecutionException(JobID jobID, String msg, Throwable cause) {
+	public JobExecutionException(JobID jobID, ApplicationStatus status, String msg, Throwable cause) {
 		super(msg, cause);
 		this.jobID = jobID;
+		this.status = status;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java:49]: Shouldn't we assert that `status` is not `SUCCEEDED`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java [hunk 8]
```diff
@@ -221,7 +222,7 @@ public static ExecutionGraph buildGraph(
 				vertex.initializeOnMaster(classLoader);
 			}
 			catch (Throwable t) {
-					throw new JobExecutionException(jobId,
+					throw new JobExecutionException(jobId, ApplicationStatus.fromJobStatus(executionGraph.getState()),
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java:225]: I think it is not correct to already decide on the `ApplicationStatus` here. If the initialization of the job fails, we might still be able to retry it. Maybe it would be better to not use the `JobExecutionException` across the client and server. Maybe it should stay a client side concept which is only thrown by the client when obtaining the job result.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/ClusterClientJobClientAdapter.java [hunk 9]
```diff
@@ -116,8 +115,7 @@ public JobID getJobID() {
 						try {
 							return jobResult.toJobExecutionResult(classLoader);
 						} catch (Throwable t) {
-							throw new CompletionException(
-									new ProgramInvocationException("Job failed", jobID, t));
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/ClusterClientJobClientAdapter.java:120]: Aren't we changing here part of the public API by changing the return type of the exceptions? How much of a contract is it that the user of the JobClient expects an `ProgramInvocationException` @aljoscha?

**Comment by aljoscha** [comment]
To me it looks good! So I think we can change it. However, it's still not optimal for users because `JobExecutionException` is not really a user-facing exception, it's defined in the `flink-runtime` package.

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating this PR @SteNicholas. In general I like the unification of thrown exception. I had two questions:

1) Is the `ProgramInvocationException` part of the public API of a `JobClient`?
2) Can we remove the `JobExecutionException` from the runtime components (specifically the `ExecutionGraphBuilder`)?

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java (hunk 5 above)
Comment (SteNicholas) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:261]: @tillrohrmann , because this change checks whether the application status is FAILED. Only when application status is FAILED, this could throw Exception with fail message.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java (hunk 6 above)
Comment (SteNicholas) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:346]: @tillrohrmann The same to above.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java (hunk 5 above)
Comment (tillrohrmann) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:261]: I think if program throws a `JobCancellationException`, then we don't set it in the `errorRef` variable. Hence, this test should not be necessary.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java (hunk 8 above)
Comment (SteNicholas) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java:225]: @tillrohrmann , what's the problem when using the `JobExecutionException` across the client and server?

**Comment by SteNicholas** [comment]
> 2\. ExecutionGraphBuilder

@tillrohrmann 
1.`ProgramInvocationException` isn't the part of the public API of a `JobClient`. Even if it is, it is wrapped into `CompletableFuture`.
2.Why the `JobExecutionException` is  removed from the runtime components (specifically the `ExecutionGraphBuilder`)? What are the adverse effects of this？

**Comment by SteNicholas** [comment]
> Thanks for creating this PR @SteNicholas. In general I like the unification of thrown exception. I had two questions:
> 
> 1. Is the `ProgramInvocationException` part of the public API of a `JobClient`?
> 2. Can we remove the `JobExecutionException` from the runtime components (specifically the `ExecutionGraphBuilder`)?

@tillrohrmann , thanks for your detailed review. My answer for above question you mentioned is as follows:
1.ProgramInvocationException isn't the part of the public API of a JobClient. Even if it is, it is wrapped into CompletableFuture.
2.Why the JobExecutionException is removed from the runtime components (specifically the ExecutionGraphBuilder)? What are the adverse effects of this？

File: flink-clients/src/main/java/org/apache/flink/client/deployment/ClusterClientJobClientAdapter.java (hunk 9 above)
Comment (SteNicholas) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/ClusterClientJobClientAdapter.java:120]: @tillrohrmann , IMO, this is about the changes to `CompletionExcetpion` contract for `JobClient `, which maybe change much. @aljoscha Right?

File: flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java (hunk 7 above)
Comment (SteNicholas) [reply flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java:49]: @tillrohrmann , no, there is a case that `jobResult.toJobExecutionResult` throws excepton with SUCCESS application status.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java (hunk 8 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java:225]: One problem is that we are specifying the "final" application status for the `JobExecutionException` here. I think this is wrong since `executionGraph.getState()` could also be something different than a globally terminal state.

File: flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java (hunk 7 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java:49]: Which case is this exactly? I thought that we would obtain a `JobExecutionResult` if the execution of the job was successful.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/ClusterClientJobClientAdapter.java (hunk 9 above)
Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/ClusterClientJobClientAdapter.java:120]: I'd like to hear the opinion of someone involved in the APIs of Flink before making any decisions here.

**Comment by tillrohrmann** [comment]
Thanks for the update @SteNicholas. Before moving forward I would like to hear from someone involved in the public APIs, that the `ProgramInvocationException` is indeed not public. Moreover, I still believe that the `JobExecutionException` which now transports the final `ApplicationStatus` shouldn't be used within the runtime components. Instead it should be used to communicate a failed `ApplicationStatus` to the outside world.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java (hunk 5 above)
Comment (SteNicholas) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:261]: @tillrohrmann , I got your point for this change, and I would like to update commit for this.

File: flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java (hunk 7 above)
Comment (SteNicholas) [reply flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java:49]: @tillrohrmann , the case is that the execution of the job is successful, but calls `AccumulatorHelper.deserializeAccumulators` throws Exception while calling `jobResult.toJobExecutionResult`.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/ClusterClientJobClientAdapter.java (hunk 9 above)
Comment (SteNicholas) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/ClusterClientJobClientAdapter.java:120]: @tillrohrmann , I would wait for @aljoscha to give the opinion of the changes to `CompletionExcetpion` contract for `JobClient`.

**Comment by SteNicholas** [comment]
Thanks for @tillrohrmann detailed review reply. I agree with the point that `JobExecutionException` transporting the final `ApplicationStatus` shouldn't be used within the runtime components. I would change the commit in `ExecutionGraphBuilder` to remove the `JobExecutionException` from `runtime ` components.
BTW, @aljoscha , could you please give your opinion about the changes to `CompletionExcetpion` contract for `JobClient`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java (hunk 7 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java:49]: Wouldn't we then fail with an `IOException` being thrown from `JobResult#toJobExecutionResult`?

Comment (SteNicholas) [reply flink-runtime/src/main/java/org/apache/flink/runtime/client/JobExecutionException.java:49]: In `MinCluster#executeJobBlocking`, if `JobResult#toJobExecutionResult` throws the `IOException`, the method would catch the `IOException` and throws the `JobExecutionException`. Meanwhile, the application status of `JobExecutionException` is `SUCCEEDED `. Right?

**Comment by kl0u** [comment]
Although the `ProgramInvocationException` is not annotated as `Public` or `PublicEvolving`, due to the fact that it has been here since forever, it seems that in some places we may be handling it specially and even expose it to the users. For example, the `ClientUtils.executeProgram()` is public and in the mailing list in the past there were people saying that they use it in other projects (I think Beam used to use it and I do not know if they still do).

On a separate note, given that this PR although small it has taken more than 3 weeks and a lot of discussions, I am starting to doubt if we should move forward with it. For example, the question that @tillrohrmann mentioned to have a separate exception for the runtime components and a separate for the client so that we can handle them differently, seems a deeper issue that requires more thought.

I may be wrong on that, but the discussions seem to be disproportionately long for the size and also the added value of this PR. What do you think? 

BTW sorry for your time @SteNicholas if this does not get merged in the end but when I opened the JIRA I did not have a clear view of all the related changes that would pop up.

**Comment by SteNicholas** [comment]
> example

@kl0u , IMO, I really want to follow the remaining comments from @tillrohrmann , and continue to update the commit according to Till's above comments. It's better for me to merge this pull request in the end.

**Comment by SteNicholas** [comment]
> Although the `ProgramInvocationException` is not annotated as `Public` or `PublicEvolving`, due to the fact that it has been here since forever, it seems that in some places we may be handling it specially and even expose it to the users. For example, the `ClientUtils.executeProgram()` is public and in the mailing list in the past there were people saying that they use it in other projects (I think Beam used to use it and I do not know if they still do).
> 
> On a separate note, given that this PR although small it has taken more than 3 weeks and a lot of discussions, I am starting to doubt if we should move forward with it. For example, the question that @tillrohrmann mentioned to have a separate exception for the runtime components and a separate for the client so that we can handle them differently, seems a deeper issue that requires more thought.
> 
> I may be wrong on that, but the discussions seem to be disproportionately long for the size and also the added value of this PR. What do you think?
> 
> BTW sorry for your time @SteNicholas if this does not get merged in the end but when I opened the JIRA I did not have a clear view of all the related changes that would pop up.

@kl0u , IMO, I really want to follow the remaining comments from @tillrohrmann , and continue to update the commit according to Till's above comments. It's better for me to merge this pull request in the end.
